### DISCUSSION TITLE
模拟器新特性讨论

### DISCUSSION THREAD

**Discussion by qiufuyu123** [Physics365/991CN-X-CW-Decompilation#18] [Ideas]
关于模拟器大家有什么新的需求/建议吗？
（991cncw支持已经完成，假期发布release版本)

**Comment by Wenti-D** [comment]
ImGui 有停靠（docking）功能，就有点像你把桌面上某个窗口拉到屏幕最右边 Windows 给你的反应。可以在模拟器后台窗口上加上这个，能更容易分配窗口空间。

**Reply by qiufuyu123** [reply]
![image](https://github.com/Physics365/991CN-X-CW-Decompilation/assets/60765480/f1e76cb1-0f97-4e2b-b168-a12dc1c0ef9b)
最新commit已经实现，可以git pull编译
二进制发布版本合并到下一次更新

**Comment by EwTE-H** [comment]
把所有的通用寄存器都搞到一行或两行里，大概像这样，一屏就能看全了
```
0  1  2  3  4  5  6
11 45 14 19 19 81 00  ..
```
还有就是，建议step的时候不要冒出来断点，而且断点执行过一次后自动消失，否则step几下就搞得到处都是断点，还删不干净
以及建议一个内存监视功能，可以在读到/写到某个地址时暂停（感觉这个比代码断点还有用，因为很多时候并不知道代码的地址）

**Reply by qiufuyu123** [reply]
step断点问题修复，见最新tag

**Comment by Physics365** [comment]
1.在内存查看/编辑窗口中使用Ctrl+V一次复制多个内存
2.上次提的分辨率和快捷键
3.（不是很需要）在_disas.txt窗口旁边放上反编译出来的C语言
4.直接修改寄存器的值

**Comment by fish4terrisa-MSDSM** [comment]
从字符串加载rop时合并大小写的16进制（可以在程序内部加一个大小写转换的说），并且自动每两个字符分割（不必手打空格）

**Reply by qiufuyu123** [reply]
已经修改
![image](https://github.com/Physics365/991CN-X-CW-Decompilation/assets/60765480/07ca6c6e-3d12-45da-8723-e8368fa6b33b)

**Comment by qiufuyu123** [comment]
https://github.com/Physics365/991CN-X-CW-Decompilation/assets/60765480/deccbf26-63c1-434b-8482-8d9e62e44e91


这个寄存器布局大家觉得怎么样？ @EwTE-H @Physics365

**Reply by Physics365** [reply]
1. 布局挺好
2. 可以把窗口名称外面的蓝框去掉
3. (已弃用)可以把模拟器也塞到窗口里，一起运行
4. RXn为什么显示的有点挤
5. 考虑放个更好的字体（比如Cascadia Code）

**Reply by Wenti-D** [reply]
> 5. 考虑放个更好的字体（比如Cascadia Code）

其实 Unifont 是我建议使用的，因为 ImGui 默认字体就是像素字体。但现在随着界面的奇怪缩放，一像素不再是一像素，所以看起来很糊。我觉得更行的中英文均支持的代码字体是[等距更纱黑体（Sarasa Mono）](https://github.com/be5invis/Sarasa-Gothic)。

**Reply by qiufuyu123** [reply]
![image](https://github.com/Physics365/991CN-X-CW-Decompilation/assets/60765480/ed3a1b50-e07c-4d0f-ad2d-6c06997f04d5)
这样舒服一点，标题去掉（

**Reply by Wenti-D** [reply]
> 3. 可以把模拟器也塞到窗口里，一起运行

之前说想要提升模拟器的像素，也就是要找到一种渲染 SVG 的方法，这样一来，模拟器可以随意放缩。但我觉得即便随意放缩也要保持比例，而如此一来把模拟器放进去就不太好分布窗口了……吧。

**Reply by Physics365** [reply]
呃 是的呢

**Reply by fish4terrisa-MSDSM** [reply]
还有，能不能在哪里的设置项里加一个是否全屏的设置。。。不知道为什么在我这边的电脑上一直默认全屏的说

**Reply by Physics365** [reply]
啊，你用的是最新版的吗

**Reply by bczhc** [reply]
> 还有，能不能在哪里的设置项里加一个是否全屏的设置。。。不知道为什么在我这边的电脑上一直默认全屏的说

是模拟器自动全屏吗？我的i3也是这样。目前可以加`resizable=0`以及配合着`width`和`height`参数，只是就不能自由缩放了

**Reply by bczhc** [reply]
这个寄存器我觉得是不是能稍微做一下定位，否则比如要找R9 R10这种数字比较大的是不是还要去数……

**Reply by bczhc** [reply]
目前字体糊我暂时使用的办法是，提升了下`io.Fonts->AddFontFromFileTTF`的pixel size，调成48之类的，然后缩放就调0.3几，字体用Noto SC Sans，2K屏上清楚爽歪歪。

**Reply by EwTE-H** [reply]
寄存器有没有LR SP EA PSW等 偶尔还是会用到的

**Reply by Wenti-D** [reply]
> 目前字体糊我暂时使用的办法是，提升了下`io.Fonts->AddFontFromFileTTF`的pixel size，调成48之类的，然后缩放就调0.3几，字体用Noto SC Sans，2K屏上清楚爽歪歪。

我去，你这一说我突然试了一下，把 `pixel size` 调成 16（Unifont 的正确尺寸），UI 缩放值为 0.75（建议查看大图）：
![image](https://github.com/Physics365/991CN-X-CW-Decompilation/assets/73067578/919ad43c-a4bc-4172-a8a7-fe721a6df8a9)

**Reply by Wenti-D** [reply]
然后这就牵扯到了 Windows 一个很奇怪的事情：明明是 16 像素的像素字体，却偏偏要乘以 3/4 才能正确一一对应。

**Reply by qiufuyu123** [reply]
![image](https://github.com/Physics365/991CN-X-CW-Decompilation/assets/60765480/de7ac1ff-a676-4aab-b121-0cde16408b3c)
最新commit已经添加这些了，等下一个release发

**Comment by Physics365** [comment]
@qiufuyu123 
那个Windows上三指下滑导致程序闪退的bug在[prerelease v0.4](https://github.com/qiufuyu123/CasioEmuNeo/releases/tag/v0.4.0)中已经修复。
可以发在release的发布说明里了。

顺便问一句：原来是哪里出了问题？

**Reply by bczhc** [reply]
可能是这个 https://github.com/qiufuyu123/CasioEmuNeo/issues/9#issuecomment-2058831490

**Comment by qiufuyu123** [comment]
内存断点初步完成 @EwTE-H 

https://github.com/Physics365/991CN-X-CW-Decompilation/assets/60765480/fae4d9b6-68a3-4c69-a1b7-45b413020210

**Comment by qiufuyu123** [comment]
## !建议更新到此版本!  
修复字体，缩放，ui性能所有已知bug  
添加内存断点功能
[release](https://github.com/qiufuyu123/CasioEmuNeo/releases/tag/v0.5.0)

**Comment by bczhc** [comment]
考虑重新加入lua控制台功能吗？有时候用lua搞些自动化还是挺方便的。比如dump出整个内存这样很基础的功能。

**Reply by Physics365** [reply]
支持

**Comment by Physics365** [comment]
考虑加入截屏功能时能渲染矢量图或高清png，像这样：
![fx-991CN X-2024_4_21 22_50_21](https://github.com/Physics365/991CN-X-CW-Decompilation/assets/154725076/f82aa862-d636-45b8-a550-18db01cc0b1e)

**Comment by Physics365** [comment]
要是模拟器能显示成下面那样就舒服多了

**Reply by Physics365** [reply]
可以改下lua的配置，放一下高清的位图。参考效果(没有防伪标)：
![fx-991CN X-2024_4_21 22_55_13](https://github.com/Physics365/991CN-X-CW-Decompilation/assets/154725076/89c23018-be82-49a4-999d-34592cb58e8c)

**Reply by Physics365** [reply]
然而我不会弄

**Comment by fish4terrisa-MSDSM** [comment]
是不是可以合并https://github.com/Xyzstk/CasioEmuX这里 @Xyzstk 大佬的修改，添加了自动按键功能，感觉好像蛮有用的

**Reply by fish4terrisa-MSDSM** [reply]
在https://github.com/Physics365/991CN-X-CW-Decompilation/discussions/28 这里提到
